XenAdmin/Commands/InstallToolsCommand.cs

```diff
@@ -132,38 +106,41 @@ private AsyncAction SingleVMExecute(VM vm)
                     var createDriveAction = new CreateCdDriveAction(vm);
 
                     using (var dlg = new ActionProgressDialog(createDriveAction, ProgressBarStyle.Marquee))
-                    {
                         dlg.ShowDialog(Parent);
-                    }
 
                     if (createDriveAction.Succeeded)
-                    {
                         ShowMustRebootBox();
-                    }
-                    return createDriveAction;
+
+                    InstallTools?.Invoke(createDriveAction);
                 }
+                return;
             }
-            else
+
+            using (var dlg = new WarningDialog(Messages.XS_TOOLS_MESSAGE_ONE_VM,
+                new ThreeButtonDialog.TBDButton(Messages.INSTALL_XENSERVER_TOOLS_BUTTON,
+                    DialogResult.OK, ThreeButtonDialog.ButtonType.ACCEPT, true),
+                ThreeButtonDialog.ButtonCancel)
             {
-                DialogResult dr = new InstallToolsWarningDialog(vm.Connection).ShowDialog(Parent);
-                if (dr == DialogResult.Yes)
+                ShowLinkLabel = true,
+                LinkText = Messages.INSTALLTOOLS_READ_MORE,
+                LinkAction = () => Help.HelpManager.Launch("InstallToolsWarningDialog")
+            })
+                if (dlg.ShowDialog(Parent) == DialogResult.OK && CheckToolSrs(vm))
                 {
                     var installToolsAction = new InstallPVToolsAction(vm, Properties.Settings.Default.ShowHiddenVMs);
                     installToolsAction.Completed += InstallToolsActionCompleted;
 
                     installToolsAction.RunAsync();
-                    return installToolsAction;
+                    InstallTools?.Invoke(installToolsAction);
                 }
-            }
-            return null;
         }
 
         /// <summary>
         /// Attempts to install tools on several VMs
         /// </summary>
         /// <param name="vms"></param>
         /// <returns>Whether the action was launched (i.e., the user didn't Cancel)</returns>

```

Comment:
This method now no longer returns anything (`void`), please update the annotation to reflect that.